autotyping/autotyping.py

```diff
@@ -513,35 +553,49 @@ def type_of_expression(expr: libcst.BaseExpression) -> Optional[Type[object]]:
     """
     if isinstance(expr, libcst.Float):
         return float
-    elif isinstance(expr, libcst.Integer):
+    if isinstance(expr, libcst.Integer):
         return int
-    elif isinstance(expr, libcst.Imaginary):
+    if isinstance(expr, libcst.Imaginary):
         return complex
-    elif isinstance(expr, libcst.FormattedString):
+    if isinstance(expr, libcst.FormattedString):
         return str  # f-strings can only be str, not bytes
-    elif isinstance(expr, libcst.SimpleString):
+    if isinstance(expr, libcst.SimpleString):
         if "b" in expr.prefix:
             return bytes
-        else:
-            return str
-    elif isinstance(expr, libcst.ConcatenatedString):
+        return str
+    if isinstance(expr, libcst.ConcatenatedString):
         left = type_of_expression(expr.left)
         right = type_of_expression(expr.right)
         if left == right:
             return left
-        else:
-            return None
-    elif isinstance(expr, libcst.Name) and expr.value in ("True", "False"):
+        return None
+    if isinstance(expr, libcst.Name) and expr.value in ("True", "False"):
         return bool
-    elif (
-        isinstance(expr, libcst.Call)
-        and isinstance(expr.func, libcst.Attribute)
-        and isinstance(expr.func.value, libcst.BaseString)
-        and expr.func.attr.value in ("format", "lower", "upper", "title")
+    if isinstance(expr, libcst.UnaryOperation) and isinstance(
+        expr.operator, libcst.Not
     ):
+        return bool
+    if isinstance(expr, libcst.Comparison) and len(expr.comparisons) == 1:
+        operator = expr.comparisons[0].operator
+        if isinstance(operator, libcst.Is):
+            return bool
+    method = get_str_method_name(expr)
+    if method in STR_STR_METHODS:
         return str
-    else:
+    if method in STR_BOOL_METHODS:
+        return bool
+    return None
+
+
+def get_str_method_name(expr: libcst.BaseExpression) -> Optional[str]:
+    """If the expression is a calling of an str method, return the method name."""
+    if not isinstance(expr, libcst.Call):
+        return None
+    if not isinstance(expr.func, libcst.Attribute):
+        return None
+    if not isinstance(expr.func.value, libcst.SimpleString):
```

Comment:
This is still wrong since `SimpleString` can include bytes strings. You can actually check here that `type_of_expression(expr.func.value) is str` to handle this correctly.